ethcore/src/miner/zero_gas_price_checker.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity. If not, see <http://www.gnu.org/licenses/>.
+
+use client::MiningBlockChainClient;
+use transaction::SignedTransaction;
+use util::{U256, Uint};
+
+const ZGP_CONTRACT_REGISTRY_NAME: &'static str = "zgp_checker";
+
+/// Zero gas price transactions checker.
+pub trait ZeroGasPriceChecker: Send {
+	/// Checks if zgp-transaction can be appended to the transaction queue.
+	fn check(&self, chain: Option<&MiningBlockChainClient>, tx: &SignedTransaction) -> Result<bool, String>;
+}
+
+/// Implementation of `ZeroGasPriceChecker`, based on on-chain deployed contract.
+pub struct OnChainZeroGasPriceChecker {
+	contract: provider::Contract,
+}
+
+impl OnChainZeroGasPriceChecker {
+	/// Try to create instance, reading contract address from given chain.
+	pub fn from_chain(chain: &MiningBlockChainClient) -> Option<Self> {
```

Comment:
i would rename `chain` to `client` since it is, in fact, a `Client`, not a `Blockchain`.